android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch

```diff
@@ -0,0 +1,1112 @@
++	};
++	struct snd_pcm *pcm;
++
++	err = snd_card_new(&(data->udev->dev), -1, NULL, THIS_MODULE, 0, &card);
++	if (err)
++		return err;
++	data->card = card;
++
++	if ((err = snd_device_new(card, SNDRV_DEV_LOWLEVEL, data, &ops)) < 0) {
++		snd_card_free(card);
++		printk("error in creating the device");
++		return err;
++	}
++	err = snd_pcm_new(card, "USB audio source", DEVICE_INDEX,
++			  NUM_PLAYBACK_SUBSTREAMS, NUM_CAPTURE_SUBSTREAMS,
++			  &pcm);
++	if (err)
++		goto pcm_fail;
++
++	pcm->private_data = data;
++	pcm->info_flags = 0;
++	data->pcm = pcm;
++	strlcpy(pcm->name, "USB Bluetooth audio", sizeof(pcm->name));
++	snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_CAPTURE, &btusb_isoc_capture_ops);
++	snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_PLAYBACK, &btusb_isoc_playback_ops);
++	err = snd_pcm_lib_preallocate_pages_for_all(pcm, SNDRV_DMA_TYPE_DEV,
++			NULL, 64 * 1024, 64 * 1024);
++	if (err) {
++		pr_emerg("snd_pcm_lib_preallocate_pages_for_all err:%d\n", err);
++		goto pcm_fail;
++	}
++	strlcpy(card->driver, "bt_audio_source", sizeof(card->driver));
++	strlcpy(card->shortname, card->driver, sizeof(card->shortname));
++	strlcpy(card->longname, "USB Bluetooth audio source",
++			sizeof(card->longname));
++
++	err = snd_card_register(card);
++	if (err){
++		goto register_fail;
++	} else {
++		printk("%s:Sound card registred sucecssfully",__func__);
++	}
++
++	dummy_proc_init(card);
++	return 0;
++
++register_fail:
++	printk("%s: Failure in	registering sound card",__func__);
++pcm_fail:
++
++	printk("%s: Failure in regitsering PCM ", __func__);
++	snd_card_free(card);
++	return err;
++}
++
++int btusb_sco_probe(struct usb_interface *intf,const struct usb_device_id *id)
++{
++	int  err = 0;
++	struct btusb_data *data;
++
++	/* data allocated during card creation */
++	data = devm_kzalloc(&intf->dev, sizeof(*data), GFP_KERNEL);
++	if (!data)
++		return -ENOMEM;
++
++	data->isoc_ifnum = intf->cur_altsetting->desc.bInterfaceNumber;
++	data->udev = interface_to_usbdev(intf);
++	data->intf = intf;
++	usb_set_intfdata(intf, data);
++
++	spin_lock_init(&data->txlock);
++	spin_lock_init(&data->rxlock);
++        init_usb_anchor(&data->rx_anchor);
++	err = btusb_snd_card_create(data);
++
++	return err;
++}
++
++
++static void btusb_sco_disconnect(struct usb_interface *intf)
++{
++	struct btusb_data *data = usb_get_intfdata(intf);
++
++
++	if (!data)
++		return;
++
++	if (data->intf)
++		usb_set_intfdata(data->intf, NULL);
++
++	snd_card_free_when_closed(data->card);
++}
++
++static int btusb_sco_ioctl(struct usb_interface *intf, unsigned int code,
++                       void *buf){
++        if(!buf)
++                return -EINVAL;
++
++        memcpy(handle,(uint8_t*)buf,SCO_HANDLE_SIZE);
++        printk("value of handle = %x, %x",handle[0],handle[1]);
++        return 0;
++}
++
++static struct usb_driver btusb_sco_driver = {
++	.name		= "btusb_sco",
++	.probe		= btusb_sco_probe,
++	.unlocked_ioctl = btusb_sco_ioctl,
++	.disconnect	= btusb_sco_disconnect,
```

Comment:
What if userspace does not call IOCTL to pass SCO handle? 

We need to handle this case. May be PCM open should fail if handle is not set properly or something else, but this case needs to be handled.